File: Other Emulators & Resources/retro-0.4/retro/PriorityQueue.java
/*
 *  PriorityQueue.java
 *  Joris van Rantwijk
 */

package retro;

/**
 * Simple heap based priority queue.
 *
 * Adding elements and removing the element with minimum priority
 * both take O(log(n)) time in most cases.  Adding an element may
 * take O(n) time only when the underlying array needs to grow, but
 * this extra cost disappears when amortized over all operations.
 * Removing a specific (non-minimum) object from the queue requires
 * sequential search and therefore takes O(n) time.
 * <p>
 * The queue is non-stable: objects with equal priority are returned
 * in arbitrary order (but the order is still deterministic).
 * <p>
 * Methods are not synchronized; external synchronization is required when
 * a priority queue is used by multiple threads.
 */
public final class PriorityQueue
{

    private int nHeap;
    private Object[] heapObj;
    private long[] heapPri;

    /** Constructs a new empty priority queue. */
    public PriorityQueue()
    {
        nHeap = 0;
        heapObj = new Object[16];
        heapPri = new long[16];
    }

    /** Removes all elements from the priority queue. */
    public void clear()
    {
        nHeap = 0;
        heapObj = new Object[16];
        heapPri = new long[16];
    }

    /** Adds an object with specified priority to the queue. */
    public void add(Object a, long prio)
    {
        nHeap++;
        if (nHeap >= heapObj.length) {
            Object[] oldHeapObj = heapObj;
            long[] oldHeapPri = heapPri;
            heapObj = new Object[2*nHeap];
            heapPri = new long[2*nHeap];
            System.arraycopy(oldHeapObj, 0, heapObj, 0, nHeap);
            System.arraycopy(oldHeapPri, 0, heapPri, 0, nHeap);
        }

        heapPri[0] = Long.MIN_VALUE;    // element 0 is a sentinel
        int k = nHeap;
        while (heapPri[k / 2] > prio) {
            heapObj[k] = heapObj[k / 2];
            heapPri[k] = heapPri[k / 2];
            k = k / 2;
        }

        heapObj[k] = a;
        heapPri[k] = prio;
    }

    /**
     * Returns the lowest priority in the queue,
     * or returns Long.MAX_VALUE if the queue is empty.
     */
    public long minPrio()
    {
        return (nHeap > 0) ? heapPri[1] : Long.MAX_VALUE;
    }

    /**
     * Removes the object with lowest priority from the queue and returns it,
     * or returns null if the queue is empty.
     */
    public Object removeFirst()
    {
        if (nHeap == 0)
            return null;

        Object a = heapObj[1];

        Object vo = heapObj[nHeap];
        long   vp = heapPri[nHeap];
        nHeap--;

        int k = 1;
        while (k <= nHeap / 2) {
            int j = 2 * k;
            if (j < nHeap && heapPri[j] > heapPri[j+1])
                j++;
            if (vp <= heapPri[j])
                break;
            heapObj[k] = heapObj[j];
            heapPri[k] = heapPri[j];
            k = j;
        }
        heapObj[k] = vo;
        heapPri[k] = vp;

        return a;
    }

    /** Removes the specified object from the queue. */
    public void remove(Object a)
    {
        int k;
        for (k = 1; k <= nHeap && heapObj[k] != a; k++) ;

        if (k <= nHeap) {
            Object vo = heapObj[nHeap];
            long   vp = heapPri[nHeap];
            nHeap--;

            while (k <= nHeap / 2) {
                int j = 2 * k;
                if (j < nHeap && heapPri[j] > heapPri[j+1])
                    j++;
                if (vp <= heapPri[j])
                    break;
                heapObj[k] = heapObj[j];
                heapPri[k] = heapPri[j];
                k = j;
            }
            heapObj[k] = vo;
            heapPri[k] = vp;
        }
    }

    /** Returns the number of elements in the queue. */
    public int size()
    {
        return nHeap;
    }

    /** Returns true if the queue does not contain any elements. */
    public boolean isEmpty()
    {
        return (nHeap == 0);
    }

}

/* end */

File: x8086NetEmu/Helpers/Misc/PriorityQueue.vb
Public Class PriorityQueue
    Private nHeap As Integer = 0
    Private heapObj(16 - 1) As Object
    Private heapPri(16 - 1) As Long

    Public Sub New()
    End Sub

    Public Sub Clear()
        nHeap = 0
        ReDim heapObj(16 - 1)
        ReDim heapPri(16 - 1)
    End Sub

    Public Sub Add(obj As Object, priority As Long)
        nHeap += 1
        If nHeap >= heapObj.Length Then
            Dim oldHeapObj() As Object = heapObj
            Dim oldHeapPri() As Long = heapPri
            ReDim heapObj(2 * nHeap - 1)
            ReDim heapPri(2 * nHeap - 1)
            Array.Copy(oldHeapObj, 0, heapObj, 0, nHeap)
            Array.Copy(oldHeapPri, 0, heapPri, 0, nHeap)
        End If

        heapPri(0) = Long.MinValue ' element 0 is a sentinel
        Dim k As Integer = nHeap
        While heapPri(k \ 2) > priority
            heapObj(k) = heapObj(k \ 2)
            heapPri(k) = heapPri(k \ 2)
            k = k \ 2
        End While

        heapObj(k) = obj
        heapPri(k) = priority
    End Sub

    Public Function MinPriority() As Long
        Return If(nHeap > 0, heapPri(1), Long.MaxValue)
    End Function

    Public Function RemoveFirst() As Object
        If nHeap = 0 Then Return Nothing

        Dim obj As Object = heapObj(1)

        Dim vo As Object = heapObj(nHeap)
        Dim vp As Long = heapPri(nHeap)
        nHeap -= 1

        Dim k As Integer = 1
        Dim j As Integer
        While k <= nHeap \ 2
            j = 2 * k
            If j < nHeap AndAlso heapPri(j) > heapPri(j + 1) Then j += 1
            If vp <= heapPri(j) Then Exit While

            heapObj(k) = heapObj(j)
            heapPri(k) = heapPri(j)
            k = j
        End While
        heapObj(k) = vo
        heapPri(k) = vp

        Return obj
    End Function

    Public Sub Remove(obj As Object)
        Dim k As Integer = 1
        While k <= nHeap AndAlso (Not heapObj(k) Is obj)
            k += 1
        End While

        If k <= nHeap Then
            Dim vo As Object = heapObj(nHeap)
            Dim vp As Long = heapPri(nHeap)
            nHeap -= 1

            Dim j As Integer
            While k <= nHeap \ 2
                j = 2 * k
                If j < nHeap AndAlso heapPri(j) > heapPri(j + 1) Then j += 1
                If vp <= heapPri(j) Then Exit While

                heapObj(k) = heapObj(j)
                heapPri(k) = heapPri(j)
                k = j
            End While
            heapObj(k) = vo
            heapPri(k) = vp
        End If
    End Sub

    Public ReadOnly Property Size As Integer
        Get
            Return nHeap
        End Get
    End Property

    Public ReadOnly Property IsEmpty As Boolean
        Get
            Return nHeap = 0
        End Get
    End Property
End Class
